### PR TITLE
FreeBSD support

### PR DESCRIPTION
**Pull request by AMDmi3**

### CODE REVIEW

**Comment by mapsmetest** [comment]
Jenkins asks: can anyone have a look at this request?
If you want Jenkins to compile and test this, write "test this please" or JT_TP (without the underscore in the middle).

**Comment by AMDmi3** [comment]
Хм, test this please

**Comment by AMDmi3** [comment]
test this please

File: routing/routing_integration_tests/routing_integration_tests.pro [hunk 1]
```diff
@@ -11,7 +11,7 @@ CONFIG -= app_bundle
 TEMPLATE = app
 
 ROOT_DIR = ../..
-DEPENDENCIES = map routing search storage indexer platform geometry coding base osrm jansson protobuf tomcrypt succinct stats_client
+DEPENDENCIES = map routing search storage indexer platform geometry coding base osrm jansson protobuf tomcrypt succinct stats_client opening_hours
```
Comment (mnemonicflow) [inline routing/routing_integration_tests/routing_integration_tests.pro]: Any reason why the opening_hours target was added to the routing integration tests?

File: common.pri [hunk 2]
```diff
@@ -61,7 +61,7 @@ RCC_DIR = $$TEMP_PATH
 MOC_DIR = $$TEMP_PATH
 UI_DIR = $$TEMP_PATH
 
-QMAKE_LIBDIR *= $$BINARIES_PATH
+LIBPATH = $$BINARIES_PATH $$LIBPATH
```
Comment (mnemonicflow) [inline common.pri]: Why have you changed the variable name? Variable LIBPATH is deprecated; use QMAKE_LIBDIR instead

File: configure.sh [hunk 3]
```diff
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/bin/sh
```
Comment (Zverik) [inline configure.sh]: Почему? Bash-а нет на freebsd? Некоторые скрипты (едва ли этот) используют спец. возможности баша.

File: routing/routing_integration_tests/routing_integration_tests.pro (hunk 1 above)
Comment (Zverik) [inline routing/routing_integration_tests/routing_integration_tests.pro]: Как-то это не в тему про freebsd?

File: platform/platform_freebsd.cpp [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+
+#include <stdlib.h>
+#include <unistd.h>
+
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <sys/sysctl.h>
+
+
+/// @return directory where binary resides, including slash at the end
```
Comment (ygorshenin) [inline platform/platform_freebsd.cpp]: We do not use static functions, we use anonymous namespaces instead.

File: platform/platform_freebsd.cpp [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+static bool GetBinaryFolder(string & outPath)
```
Comment (ygorshenin) [inline platform/platform_freebsd.cpp]: s/Folder/Dir/g

File: platform/platform_freebsd.cpp [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+    outPath = path;
+    outPath.erase(outPath.find_last_of('/') + 1);
+    return true;
+  }
+  return false;
+}
+
+Platform::Platform()
+{
+  // init directories
```
Comment (ygorshenin) [inline platform/platform_freebsd.cpp]: All comments except inline should start from a capital letter and end with a dot.

File: platform/platform_freebsd.cpp [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+  string path;
+  CHECK(GetBinaryFolder(path), ("Can't retrieve path to executable"));
+
+  char const * homePath = ::getenv("HOME");
+  string const home(homePath ? homePath : "");
+
+  m_settingsDir = home + "/.config/MapsWithMe";
```
Comment (ygorshenin) [inline platform/platform_freebsd.cpp]: Please, use JoinFoldersToPath() function.

File: platform/platform_freebsd.cpp [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+
+  if (!IsFileExistsByFullPath(m_settingsDir + SETTINGS_FILE_NAME))
+  {
+    mkdir((home + "/.config/").c_str(), 0755);
```
Comment (ygorshenin) [inline platform/platform_freebsd.cpp]: Why not to use Platform::MkDir()

File: platform/platform_freebsd.cpp [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+    mkdir(m_settingsDir.c_str(), 0755);
+  }
+
+  m_writableDir = home + "/.local/share/MapsWithMe";
+  mkdir((home + "/.local/").c_str(), 0755);
+  mkdir((home + "/.local/share/").c_str(), 0755);
+  mkdir(m_writableDir.c_str(), 0755);
+
+  char const * resDir = ::getenv("MWM_RESOURCES_DIR");
+  if (resDir)
```
Comment (ygorshenin) [inline platform/platform_freebsd.cpp]: This function looks like platform/platform_linux.cpp. It's not good to just duplicate it's code here. Please, reuse old code instead of copy-and-paste.

File: platform/platform_freebsd.cpp [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+    return content.substr(0, 32);
+  }
+  else
+    return "n0dbus0n0hostid00000000000000000";
+
+}
+
+void Platform::RunOnGuiThread(TFunctor const & fn)
+{
+  /// @todo
```
Comment (ygorshenin) [inline platform/platform_freebsd.cpp]: Who will fix that @todo?

File: platform/platform_freebsd.cpp [hunk 11]
```diff
@@ -0,0 +1,142 @@
+#include "platform/platform.hpp"
```
Comment (Zverik) [inline platform/platform_freebsd.cpp]: Насколько сложно вписать это в platform_linux, насыпав `#ifdef`?

Comment (ygorshenin) [inline platform/platform_freebsd.cpp]: The correct way is to create a separate platform_desktop_unix.{cpp|hpp} where all common Linux/FreeBSD code should reside. Linux is not FreeBSD and FreeBSD is not Linux. All Linux/FreeBSD-specific methods should be in platform_linux.cpp and platform_freebsd.cpp correspondingly.

**Comment by therearesomewhocallmetim** [comment]
Please, note that currently we have __no__ FreeBSD servers, and in order to test this, we will have to install one.

File: common.pri (hunk 2 above)
Comment (AMDmi3) [inline common.pri]: Ok

File: configure.sh (hunk 3 above)
Comment (AMDmi3) [inline configure.sh]: No, there's no bash on FreeBSD by default, and making a script that doesn't need it depend on it is a bad practice.

File: routing/routing_integration_tests/routing_integration_tests.pro (hunk 1 above)
Comment (AMDmi3) [inline routing/routing_integration_tests/routing_integration_tests.pro]: Will split to separate PR.

File: platform/platform_freebsd.cpp (hunk 5 above)
Comment (AMDmi3) [inline platform/platform_freebsd.cpp]: (replying to random comment)

This file is based on platform_linux.cpp, and it actually feels quite strange to be asked for style fixes in the code I haven't touched and to follow style upstream code doesn't. Will use common linux/freebsd code and fix bits that can be fixes though.

Comment (ygorshenin) [inline platform/platform_freebsd.cpp]: platform_linux.cpp was written before style guide was established, and, of course, it should be rewritten in accordance with our modern practices. But I can't approve *new* code (platform_freebsd.cpp) which follows old and wrong style.

Comment (Zverik) [inline platform/platform_freebsd.cpp]: So basically, to get this PR accepted, you should make your changes in `platform_linux.cpp` instread.

Comment (ygorshenin) [inline platform/platform_freebsd.cpp]: Yep, that's right. I'll fix it ASAP and notify you when it'll be fixed.

Comment (AMDmi3) [inline platform/platform_freebsd.cpp]: Thank you!

File: platform/platform_freebsd.cpp (hunk 10 above)
Comment (AMDmi3) [inline platform/platform_freebsd.cpp]: No idea. I don't see why it should be different from Linux, and I don't know why it hasn't been implemented for Linux. I guess I'll leave it in unix-generic code.

Comment (ygorshenin) [inline platform/platform_freebsd.cpp]: Agree, these functions will go to unix-generic code.

File: platform/platform_freebsd.cpp (hunk 11 above)
Comment (deathbaba) [inline platform/platform_freebsd.cpp]: The better way can be to refactor our Platform class, probably by returning to initial virtual class methods solution and their implementation for each platform. Current approach is faster at runtime but very hard to understand and support.

Comment (AMDmi3) [inline platform/platform_freebsd.cpp]: :+1:, virtual base class would be much easier to handle. Now, as all stuff is initialized in ctor, you can't basically use virtual methods, so you have to parametrize base class ctor with e.g. executable path, which is non-flexible.

Comment (deathbaba) [inline platform/platform_freebsd.cpp]: Exactly. We used virtual base class in 2010, then moved to current solution, but now I see that it was bad idea.

I think it would be great to refactor whole platform to use more clear virtual base class, rather than patch existing code (and people who port for other platforms will be grateful too). What do you think, is it feasible for you? (in a separate request, of course).

Comment (AMDmi3) [inline platform/platform_freebsd.cpp]: Sure, I can try.

Comment (mpimenov) [inline platform/platform_freebsd.cpp]: Some say that even Linux is not Linux.
It's GNU/Linux.

File: platform/platform_freebsd.cpp [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+  LOG(LDEBUG, ("Tmp directory:", m_tmpDir));
+  LOG(LDEBUG, ("Settings directory:", m_settingsDir));
+  LOG(LDEBUG, ("Client ID:", UniqueClientId()));
+}
+
+string Platform::UniqueClientId() const
+{
+  bool filterdashes = false;
+  string machineFile = "/var/db/dbus/machine-id";
+  if (IsFileExistsByFullPath("/etc/hostid")) {
```
Comment (mpimenov) [inline platform/platform_freebsd.cpp]:
```
if (IsFileExistsByFullPath("/etc/hostid")) 
{
}
```

File: platform/platform_freebsd.cpp (hunk 11 above)
Comment (ygorshenin) [inline platform/platform_freebsd.cpp]: @deathbaba , why do we need virtuals at all? There is always one and only one implementation of Platform on each platform. It doesn't make sense to compile PlatformWin on Android, for example. We have been used separate compilation for Platform instead of virtual inheritance and it works fine. From my point of view virtuals are not better that current approach, because you'll get a bunch of PlatformWin, PlatformLinux, PlatformPosix, PlatformUnix, ... classes instead of a bunch of source files.

Comment (AMDmi3) [inline platform/platform_freebsd.cpp]: You'll still have the same bunch of source files and Platform classes (though it's better to use different names and typedef used one to Platform) - there's no need to use run-time polymorphism here as there's only a single Platform implementation at a time. But virtual class will give you have strictly defined interface (you can forget to fill a variable in current implementation; you can't forget to implement virtual method) and an ability to keep common code in intermediate classes. My idea is

```
// platform.h
class IPlatform {
   virtual string GetExecutableName() = 0;
   virtual string GetSettingsDir() = 0;
};
```
```
// platform_unix.h
class UnixPlatform : public IPlatform {
    virtual string GetSettingsDir() override { return getenv("HOME") + "./mapsme"; }
};
```
```
// platform_freebsd.h
class FreeBSDPlatform : public UnixPlatform {
    virtual string GetExecutablePath() final { return ... }
};

typedef FreeBSDPlatform Platform; // all consumers work with Platform -
                                  // a final class, no runtime polymorphism
```

Comment (ygorshenin) [inline platform/platform_freebsd.cpp]: OK, that makes sense.

Comment (deathbaba) [inline platform/platform_freebsd.cpp]: @ygorshenin Virtual classes will make code much simpler and easier to support. And it’s very hard to customize member variables now for Platform (check Android implementation for example). I agree that we don’t need runtime polymorphism, but it’s just lesser evil.
And idea with defines to avoid unnecessary virtual calls sounds interesting :)

Comment (AMDmi3) [inline platform/platform_freebsd.cpp]: > I agree that we don’t need runtime polymorphism, but it’s just lesser evil.

By

> there's no need to use run-time polymorphism here

I've meant that there will be no runtime polymorphism in fact. With C++11 override & final keywords properly used, when final subclass is used directly (our case), compilers can do devirtualization and avoid vtbl lookups completely. There will not be any noticeable effect on performance here anyway.

Comment (deathbaba) [inline platform/platform_freebsd.cpp]: Great! Didn’t know about that kind of optimization, thanks.

**Comment by deathbaba** [comment]
@AMDmi3 Any news? Should we close this one?

**Comment by deathbaba** [comment]
Closing request as we discussed, that Platform should be refactored first to have base virtual class.
